src/bus.rs

```
@@ -305,6 +311,83 @@ where
 	}
 }
 
+#[derive(Debug)]
+pub struct Response<T> {
+	pub motor_id: u8,
+	pub alert: bool,
+	pub data: T,
+}
+
+impl<'a, ReadBuffer, WriteBuffer> From<StatusPacket<'a, ReadBuffer, WriteBuffer>> for Response<()>
+where
+	ReadBuffer: AsRef<[u8]> + AsMut<[u8]>,
+	WriteBuffer: AsRef<[u8]> + AsMut<[u8]>,
+{
+	fn from(status_packet: StatusPacket<'a, ReadBuffer, WriteBuffer>) -> Self {
+		Self {
+			data: (),
+			motor_id: status_packet.packet_id(),
+			alert: status_packet.alert(),
+		}
+	}
+}
+
+impl<'a, ReadBuffer, WriteBuffer> From<StatusPacket<'a, ReadBuffer, WriteBuffer>> for Response<Vec<u8>>
+where
+	ReadBuffer: AsRef<[u8]> + AsMut<[u8]>,
+	WriteBuffer: AsRef<[u8]> + AsMut<[u8]>,
+{
+	fn from(status_packet: StatusPacket<'a, ReadBuffer, WriteBuffer>) -> Self {
+		Self {
+			data: status_packet.parameters().to_owned(),
+			motor_id: status_packet.packet_id(),
+			alert: status_packet.alert(),
+		}
+	}
+}
+
+impl<'a, ReadBuffer, WriteBuffer> From<StatusPacket<'a, ReadBuffer, WriteBuffer>> for Response<u8>
```

Hmm, we could use the `TryFrom` trait so we can report an error if the data is the wrong length.

Before, the check and the code that would panic on wrong lengths were close together. But now the trait impl is far from the check that prevents runtime panics.

I like the idea of using a conversion trait, but I think `TryFrom` is more appropriate.

---

Yes, I can see them now :D 

You're right that these wont panic in the places where they are used now. But `StatusPacket` and `Response` are public types, so the user can manually call the `From` implementation too.

Besides that I also think it's nicer when the checks and the otherwise panicking code are close together. It's much easier to verify that the assumptions of the conversion are checked then.

---

> Besides that I also think it's nicer when the checks and the otherwise panicking code are close together. It's much easier to verify that the assumptions of the conversion are checked then.

Does this mean we shouldn't use the From or TryFrom traits at all? As it hides the conversion and potential resulting error. 
Or I should change the From to TryFrom and return a result type with an appropriate error. 
Just clarifying 😊

---

TryFrom should be fine, because then we can put the relevant checks inside the TryFrom implementation and report errors :)